mm2src/common/number_type_casting.rs

```diff
@@ -0,0 +1,125 @@
+/// This module contains type casting functions for numbers in a
+/// safe way.
+///
+/// The problem comes with when casting a type that supports higher value
+/// than the target type's highest value.
+///
+/// eg:
+/// ```rs
+/// let x: u64 = 4294967295 + 10;
+/// assert_eq!(x as u32, std::u32::MAX, "{} is not {}", x as u32, std::u32::MAX);
+/// ```
+
+pub trait SafeTypeCastingNumbers<T>: Sized {
+    fn into_or(self, or: T) -> T;
+    fn into_or_max(self) -> T;
+}
+
+macro_rules! impl_safe_number_type_cast {
+    ($from: ident, $to: ident) => {
+        impl SafeTypeCastingNumbers<$to> for $from {
+            fn into_or(self, or: $to) -> $to { std::convert::TryFrom::try_from(self).unwrap_or(or) }
+            fn into_or_max(self) -> $to { std::convert::TryFrom::try_from(self).unwrap_or(std::$to::MAX) }
```

Comment:
I find `into_or_max` a bit confusing for `signed -> unsigned` cast like `impl_safe_number_type_cast!(i128, u64);`. IMHO, it won't have a realistic use case. For example, if we get a negative value for a number supposed to represent an amount of money (e.g. BTC satoshis), we would likely prefer to return an error in such case instead of using max possible value. In general, should we maybe consider using `TryFrom/TryInto` directly and use `unwrap_or` where applicable?

Reply:
> I find into_or_max a bit confusing for signed -> unsigned cast like

`impl_safe_number_type_cast!(i128, u64);` so we will be able to use `into_or_zero` or `into_or`, there we should not use `into_or_max`.

> For example, if we get a negative value for a number supposed to represent an amount of money (e.g. BTC satoshis), we would likely prefer to return an error in such case instead of using max possible value.

Then if we use this implementation, we should also add function that takes closure as an argument and uses it if `try_from` fails.

> In general, should we maybe consider using TryFrom/TryInto directly and use unwrap_or where applicable?

For me it's also fine. Anything that replaces wrong `x as y` castings is fine enough :)

Reply:
> I changed my point of view, so I agree with Artem about this. It's better to handle negative values manually.

I guess we should close the PR then?

Reply:
No, we still need to cast `unsigned` to `signed` in my opinion

Reply:
@artemii235 any comments? If you aggree with Sergey, I can do that.

Reply:
Yes, I agree - let's keep `unsigned -> signed` cast and handle negative values manually.

Reply:
What about unsigned -> unsigned casting(like u64 to u16 might cause overflow and produce random value) ? @artemii235 @sergeyboyko0791

Reply:
I think it'll be helpful!
> What about unsigned -> unsigned casting

Reply:
Yeah, having e.g. `into_or_max` seems reasonable for such cases.